torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1161,23 +1161,45 @@ def clip_grad_norm_(
             self._streams["unshard"],
             self._streams["pre_unshard"],
         )
-
```

Comment:
For NO_SHARD, wondering whether we can just call the default PyTorch clip_grad_norm_() and early return without all_reduce and following logics?

Reply:
Yes, we could, but this version enables us to mix and match `NO_SHARD` and `FULL_SHARD` for different submodules. If the user knows the entire FSDP instance is only `NO_SHARD`, then they can just use `torch.nn.utils.clip_grad_norm_()`.

I think some of this logic in `FSDP.clip_grad_norm_()` will be useful when thinking about how to write `clip_grad_norm_()` for our composable APIs.

Reply:
We can optionally do a check over all FSDP instances in the module tree, and if all are `NO_SHARD`, then we can return `nn.utils.clip_grad_norm_()` directly as you suggested. Maybe I can include that fast path in a follow-up.

Reply:
I see, that makes sense!